internal/test/testdata/examples/log.wat

```diff
@@ -228,119 +228,173 @@
+
+  ;; log_header_fields logs each header field, using the given function table index.
+  (func $log_header_fields
+    (param $name i32) (param  $name_len i32)
+    (param  $buf i32) (param $buf_limit i32)
+    (param   $fn i32)
+
+    ;; result is the length of all NUL-terminated values.
+    (local $result i32)
+
+    ;; buf_log is where the log function can begin writing.
+    (local $buf_log i32)
+
+    ;; value is the position of the current name.
+    (local $value i32)
+
+    ;; len is the length of the current NUL-terminated value, exclusive of NUL.
     (local $len i32)
 
-    ;; result = table[header_fn](mem[name:name_len], mem[buf:buf_limit])
-    (local.set $result (call_indirect (type $get_header)
+    ;; result = table[headers_fn](mem[name:name_len], mem[buf:buf_limit])
+    (local.set $result (call_indirect (type $get_headers)
       (local.get $name) (local.get $name_len)
-      (local.get $buf) (local.get $buf_limit)
-      (local.get $header_fn)))
+      (local.get $buf)  (local.get $buf_limit)
+      (local.get $fn)))
 
-    ;; if result == 0 { panic }
-    (if (i64.eqz (local.get $result))
+    ;; if len == 0 { panic }
+    (if (i32.eqz (local.get $result))
        (then (unreachable))) ;; header didn't exist
 
-    ;; len = uint32(result)
-    (local.set $len (i32.wrap_i64 (local.get $result)))
-
-    ;; if len > buf_limit { panic }
-    (if (i32.gt_u (local.get $len) (local.get $buf_limit))
+    ;; if result > buf_limit { panic }
+    (if (i32.gt_u (local.get $result) (local.get $buf_limit))
        (then (unreachable))) ;; too big so wasn't written
 
-    (local.get $len))
+    ;; buf_log = buf + result
+    (local.set $buf_log (i32.add (local.get $buf) (local.get $result)))
+
+    ;; loop while we can read a NUL-terminated value.
+    (loop $values
+      ;; if mem[buf] == NUL
+      (if (i32.eqz (i32.load8_u (local.get $buf)))
+        (then ;; reached the end of the value
+
+          ;; value = buf - len
+          (local.set $value (i32.sub (local.get $buf) (local.get $len)))
+
+          ;; log_header_field(name, value, buf_log)
+          (call $log_header_field
+            (local.get $name) (local.get $name_len)
+            (local.get $value) (local.get $len)
+            (local.get $buf_log))
+
+          (local.set $buf (i32.add (local.get $buf) (i32.const 1))) ;; buf++
+          (local.set $len (i32.const 0))) ;; len = 0
+        (else
+          (local.set $len (i32.add (local.get $len) (i32.const 1))) ;; len++
+          (local.set $buf (i32.add (local.get $buf) (i32.const 1))))) ;; buf++
+
+      (local.set $result (i32.sub (local.get $result) (i32.const 1))) ;; result--
+
+      ;; if result > 0 { continue } else { break }
+      (br_if $values (i32.gt_u (local.get $result) (i32.const 0)))))
+
+  ;; log_header_field logs a header field formatted like "name: value".
+  ;;
+  ;; Note: This doesn't enforce a buf_limit as the runtime will panic on OOM.
```

Comment:
ps my general thinking is to always pass limit explicitly when making host function calls, but if you are just concatenating things already in memory, punting to let runtime panic on OOM could be fine.